File: packages/swift-code-block/Sources/SwiftCodeBlock/Render/Marks.swift
enum Marks {
    static func make(code: String, start: Int) -> [Int] {
        let bytes = code.utf8
        var count = 1
        var index = bytes.startIndex

        while index != bytes.endIndex {
            switch bytes[index] {
            case 0x0d:
                let next = bytes.index(after: index)
                if next != bytes.endIndex, bytes[next] == 0x0a {
                    index = next
                }
                count += 1
            case 0x0a:
                count += 1
            default:
                break
            }
            index = bytes.index(after: index)
        }

        let highestFirst = Int.max - (count - 1)
        let first = min(max(start, 1), highestFirst)
        return (0..<count).map { first + $0 }
    }
}

File: packages/compose-code-block/src/main/kotlin/io/github/dongyuzhao/composecodeblock/Marks.kt
package io.github.dongyuzhao.composecodeblock

internal fun marks(code: String, start: Int): List<Int> {
    var count = 1
    var index = 0
    while (index < code.length) {
        if (code[index] == '\r') {
            if (index + 1 < code.length && code[index + 1] == '\n') {
                index++
            }
            count++
        } else if (code[index] == '\n') {
            count++
        }
        index++
    }

    val lastOffset = count - 1
    val first = start.coerceIn(1, Int.MAX_VALUE - lastOffset)
    return List(count) { first + it }
}
